app/models/exercise_map.rb

```diff
@@ -0,0 +1,45 @@
+class ExerciseMap
+  def initialize(user, track)
+    @user = user
+    @track = track
+  end
+
+  def core_exercises
+    map.select(&:core?)
+  end
+
+  private
+  attr_reader :user, :track
+
+  def map
+    @map ||= exercises.map do |e|
+      ExerciseMapNode.new(
+        exercise: e,
+        unlocks: exercises_to_unlock_for(e)
+      )
+    end
+  end
+
+  def exercises
+    @exercises ||= track.
+      exercises.
+      map { |e| ExerciseWithSolution.new(e, solution_for(e)) }
+  end
+
+  def exercises_to_unlock_for(exercise)
+    exercise.unlocks.map do |exercise_to_unlock|
```

Comment:
I think this is where we also need a line that says "if this exercise auto approves then all the side exercises with unlocked_by: false" go here.